pkg/controller/plan/adapter/ovirt/client.go

```diff
@@ -460,89 +474,76 @@ func (r Client) removePrecopies(precopies []planapi.Precopy, vmService *ovirtsdk
-					} else {
-						// Try to remove the snapshot again
-						_, err = snapService.Remove().Query("correlation_id", correlationID).Send()
-						if err != nil {
-							err = liberr.Wrap(err)
-							r.Log.Error(err, "Error removing snapshot")
-							return true, err
-						}
-					}
-				}
-				return false, nil
-			})
-			if err != nil {
-				err = liberr.Wrap(err)
-				r.Log.Error(err, "Error waiting for snapshot removal")
-				return
-			}
-
-		}
-
 		for {
-			jobs, err := r.getJobs(correlationID)
-			if err != nil {
-				err = liberr.Wrap(err)
-				r.Log.Error(err, "Error waiting for snapshot removal")
-				break
-			}
-			if allJobsFinished(jobs) {
-				break
-			}
-
 			select {
 			case <-time.After(time.Duration(settings.Settings.Migration.SnapshotRemovalTimeout) * time.Minute):
 				r.Log.Info("Timeout waiting for snapshot removal")
 				return
 			default:
 				time.Sleep(time.Duration(settings.Settings.Migration.SnapshotStatusCheckRate) * time.Second)
 			}
+			// Try to remove the snapshot
+			_, err := snapService.Remove().Query("correlation_id", correlationID).Send()
+			if err != nil {
+				if strings.Contains(err.Error(), "Cannot remove Snapshot") {
+					err = web.ConflictError{
+						Provider: r.Source.Provider,
+						Err:      err,
+					}
+					r.Log.Error(err, "ConflictError failed to remove snapshot", "snapshotID", snapshotID)
+				} else {
+					err = liberr.Wrap(err)
+					r.Log.Error(err, "Remove request for snapshot failed", "snapshotID", snapshotID)
+
+					snapshotDisksResponse, err := snapService.DisksService().List().Send()
+					if err != nil {
+						r.Log.Error(err, "Request for snapshotDisksResponse failed")
+						break
+					}
+					snapshotDisks, found := snapshotDisksResponse.Disks()
+					if !found {
+						r.Log.Error(err, "Disks not found")
+						continue
+					}
+					for _, disk := range snapshotDisks.Slice() {
+						if transferred, err := r.isDiskBeingTransferred(disk.MustId()); transferred {
+							r.Log.Info("Disk is being transferred, retrying...", "disk", disk)
+						} else if err != nil {
+							err = liberr.Wrap(err)
+							r.Log.Error(err, "Error checking if disk is being transferred")
+						}
+					}
+				}
+			} else {
+				// Check the events of the snapshot
+				finished, err := r.isSnapshotRemovalFinished(correlationID)
+				if err != nil {
+					err = liberr.Wrap(err)
+					r.Log.Error(err, "Error waiting for snapshot removal")
+				}
+				if finished {
+					break
+				}
+			}
 		}
 	}
 }
 
-func allJobsFinished(jobs []*ovirtsdk.Job) bool {
-	for _, job := range jobs {
-		status, _ := job.Status()
-		if status != ovirtsdk.JOBSTATUS_FINISHED && status != ovirtsdk.JOBSTATUS_FAILED {
-			return false
+func (r *Client) isSnapshotRemovalFinished(correlationID string) (finished bool, err error) {
+	events, err := r.getEvents(correlationID)
+	if err != nil {
+		return
+	}
+	for _, event := range events {
+		switch event.MustCode() {
+		case REMOVE_SNAPSHOT_FINISHED_FAILURE:
```

Comment:
and here again, we cannot see REMOVE_SNAPSHOT_FINISHED_FAILURE that was generated for a previous attempt since if a previous attempt failed, we would have stopped trying to remove that snapshot, right?